File: Chapter13/C#/ArraySearchAlgorithms.cs
using System;
namespace EverydayDataStructures
{
	public class ArraySearchAlgorithms
	{
		public ArraySearchAlgorithms()
		{
		}

		//Sequential / Linear Search
		public int LinearSearchIndex(int[] values, int key)
		{
			for (int i = 0; i < values.Length - 1; i++)
			{
				if (values[i] == key)
				{
					return i;
				}
			}

			return -1;
		}

		public Customer LinearSearchCustomer(Customer[] customers, int custId)
		{
			for (int i = 0; i < customers.Length - 1; i++)
			{
				if (customers[i].customerId == custId)
				{
					return customers[i];
				}
			}

			return null;
		}

		//Binary Search
		public int BinarySearch(int[] values, int left, int right, int key)
		{
			if (right >= left)
			{
				int mid = left + (right - left) / 2;

				if (values[mid] == key)
				{
					return mid;
				}
				else if (values[mid] > key)
				{
					return BinarySearch(values, left, mid - 1, key);
				}

				return BinarySearch(values, mid + 1, right, key);
			}

			return -1;
		}
			
		//Jump Search
		public int JumpSearch(int[] values, int key)
		{
			int n = values.Length;
			int step = (int)Math.Sqrt(n);
			int prev = 0;

			while (values[Math.Min(step, n) - 1] < key)
			{
				prev = step;
				step += (int)Math.Floor(Math.Sqrt(n));
				if (prev >= n)
				{
					return -1;
				}
			}

			while (values[prev] < key)
			{
				prev++;
				if (prev == Math.Min(step, n))
				{
					return -1;
				}
			}
				
			if (values[prev] == key)
			{
				return prev;
			}

			return -1;
		}
	}
}

File: Chapter13/Swift/ArraySearchAlgorithms.swift
//
//  ArraySearchAlgorithms.swift
//  EverydayDataStructures-Swift
//
//  Created by William Smith on 10/31/16.
//  Copyright © 2016 Websmiths, LLC. All rights reserved.
//

import Foundation

class ArraySearchAlgorithms
{
    //Sequential / Linear
    open func linearSearhIndex( values: [Int], key: Int) -> Int
    {
        for i in 0..<values.count
        {
            if (values[i] == key)
            {
                return i
            }
        }

        return -1
    }

    open func linearSearchCustomer( customers: [Customer], custId: Int) -> Customer?
    {
        for i in 0..<customers.count
        {
            if (customers[i].custId == custId)
            {
                return customers[i]
            }
        }

        return nil
    }
    
    //Binary Search
    open func binarySearch( values: [Int], left: Int, right: Int, key: Int) -> Int
    {
        if (right >= left)
        {
            let mid: Int = left + (right - left) / 2

            if (values[mid] == key)
            {
                return mid
            }
            else if (values[mid] > key)
            {
                return binarySearch(values: values, left: left, right: mid - 1, key: key)
            }

            return binarySearch(values: values, left: mid + 1, right: right, key: key)
        }

        return -1
    }
    
    
    //Jump Search
    open func jumpSearch( values: [Int], key: Int) -> Int
    {
        let n: Int = values.count
        var step: Int = Int(sqrt(Double(n)))
        
        var prev: Int = 0
        
        while values[min(step, n) - 1] < key
        {
            prev = step
            step = step + Int(floor(sqrt(Double(n))))
            if (prev >= n)
            {
                return -1
            }
        }
        
        while (values[prev] < key)
        {
            prev = prev + 1
            if (prev == min(step, n))
            {
                return -1
            }
        }
        
        if (values[prev] == key)
        {
            return prev
        }
        
        return -1
    }
}
